### PR TITLE
Add ImageTextToText pipeline

### PR DESCRIPTION
**Pull request by NielsRogge**
# What does this PR do?

This PR adds a deprecation warning when users pass a text prompt to the `image-to-text` pipeline, and recommends to leverage a new `image-text-to-text` pipeline instead.

This way, we keep `image-to-text` for only that said task, which means image as input and text as output. Example tasks include image captioning or optical character recognition (OCR). 

The new image-text-to-text pipeline assumes input = image + text, output = text.

## Usage

Usage is as follows:

```python
from transformers import pipeline


# OK:
# model_id = "microsoft/git-base-coco"
model_id = "Salesforce/blip-image-captioning-base"
# model_id = "Salesforce/blip2-opt-2.7b" ok, although it doesn't include the text prompt in the output
# model_id = "Salesforce/instructblip-flan-t5-xl" ok, although it doesn't include the text prompt in the output
# model_id = "llava-hf/llava-1.5-7b-hf"
# model_id = "adept/fuyu-8b"
model_id = "google/pix2struct-textcaps-base"
# model_id = "microsoft/udop-large"
# model_id = "naver-clova-ix/donut-base-finetuned-docvqa"
# model_id = "microsoft/kosmos-2-patch14-224"

pipe = pipeline(task="image-text-to-text", model=model_id)

outputs = pipe(
    images="http://images.cocodataset.org/val2017/000000039769.jpg",
    text="USER: <image>\nWhat does the label 15 represent? (1) lava (2) core (3) tunnel (4) ash cloud\nASSISTANT:",
    max_new_tokens=200,
)

print(outputs)
```

## Supported models

This has been tested on:

- [x] GIT
- [x] BLIP
- [x] BLIP-2
- [x] IDEFICS
- [x] InstructBLIP
- [x] LLaVa
- [x] Fuyu
- [x] Pix2Struct
- [x] UDOP
- [x] Donut
- [x] KOSMOS-2

### CODE REVIEW

**Comment by merveenoyan** [comment]
I just tested it, thanks a lot for the PR!

**Comment by merveenoyan** [comment]
The pipeline is returning weird output for KOSMOS-2 and Fuyu-8B (and raises an error for Llava) and the output doesn't seem to be postprocessed.
```python
from transformers import pipeline
import torch 

model_id = "microsoft/kosmos-2-patch14-224"
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
pipe = pipeline(task="image-text-to-text", model=model_id, device=device)

outputs = pipe(
    images="http://images.cocodataset.org/val2017/000000039769.jpg",
    text="<grounding> Explain what the cats in the image are doing.",
    max_new_tokens=40
)

print(outputs)
# [{'generated_text': '<image>. the, to and of as in I that\' for is was- on' it with The as at bet he have from by are " you his " this said not has an ( but had we her they will my or were their): up about out who one all been she can more would It</image><grounding> Explain what the cats in the image are doing.<phrase> The two cats</phrase><object><patch_index_0049><patch_index_0799></delimiter_of_multi_objects/><patch_index_0096><patch_index_1007></object> are laying on a pink blanket, sleeping next to each other. One cat is on its back, while the other is laying on its side'}]
```

I can reproduce it with Fuyu-8B as well, I feel like something around postprocessing might be causing it. 
```python
from transformers import pipeline
import torch 

model_id = "adept/fuyu-8b"
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
pipe = pipeline(task="image-text-to-text", model=model_id, device=device)

outputs = pipe(
    images="http://images.cocodataset.org/val2017/000000039769.jpg",
    text="<grounding> Explain what the cats in the image are doing.",
    max_new_tokens=40
)

print(outputs)

# [{'generated_text': '|SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||NEWLINE||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||NEWLINE||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||NEWLINE||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||NEWLINE||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||NEWLINE||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||NEWLINE||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||NEWLINE||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||NEWLINE||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||NEWLINE||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||NEWLINE||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||NEWLINE||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||NEWLINE||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||NEWLINE||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||NEWLINE||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||NEWLINE||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||SPEAKER||NEWLINE|<s> <grounding> Explain what the cats in the image are doing.\x04 In the image, two cats are lying on a pink blanket, seemingly sleeping or resting.\n'}]
```

Llava-1.5B raises an error. 
```
from transformers import pipeline
import torch 

device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
mm_pipeline = pipeline("image-text-to-text", model="llava-hf/llava-1.5-7b-hf", device=device)

mm_pipeline("https://huggingface.co/spaces/llava-hf/llava-4bit/resolve/main/examples/baklava.png", text="How to make this pastry?")
```
`ValueError: The input provided to the model are wrong. The number of image tokens is 0 while the number of image given to the model is 1. This prevents correct indexing and breaks batch generation.`

File: src/transformers/pipelines/image_text_to_text.py [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+        records = []
+        for output_ids in model_outputs:
+            record = {
+                "generated_text": self.processor.decode(
+                    output_ids,
+                    skip_special_tokens=True,
+                )
+            }
+            records.append(record)
+        return records
```
Comment (merveenoyan) [inline src/transformers/pipelines/image_text_to_text.py:165]: For KOSMOS-2 for instance there's an additional step for postprocessing like following:
`processed_text = processor.post_process_generation(generated_text, cleanup_and_extract=False)`
Which may be why the error occurs, I'll try to see if it's the reason why

Comment (merveenoyan) [reply src/transformers/pipelines/image_text_to_text.py:165]: Ok this for sure fixes it. I'll make the necessary suggestion in the PR

File: src/transformers/pipelines/image_text_to_text.py [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+    def postprocess(self, model_outputs):
+        records = []
+        for output_ids in model_outputs:
+            record = {
+                "generated_text": self.processor.decode(
+                    output_ids,
+                    skip_special_tokens=True,
+                )
+            }
+            records.append(record)
```
Comment (merveenoyan) [inline src/transformers/pipelines/image_text_to_text.py:164]:
```suggestion
            if callable(processor.__getattribute__("post_process_generation")):
            records = self.processor.post_process_generation(record["generated_text"], cleanup_and_extract=False)

            records.append(record)
```

Something like this should work, modify as you wish

File: src/transformers/pipelines/image_text_to_text.py (hunk 1 above)
Comment (NielsRogge) [reply src/transformers/pipelines/image_text_to_text.py:144]: Yes I tried it with KOSMOS-2, and for now I wouldn't add this specific method to the pipeline, as only KOSMOS-2 has it. For now it's recommended for users to call this `post_process_generation` method themselves after getting the output of the pipeline.

In the future, I'm considering having a general `post_process_generation` method for multimodal models which require additional postprocessing after batch_decode. cc @ydshieh

Comment (ydshieh) [reply src/transformers/pipelines/image_text_to_text.py:165]: Is Kosmos-2 still the only model whose generation's text would containing tags (like bounding box info. as string)?

Comment (NielsRogge) [reply src/transformers/pipelines/image_text_to_text.py:170]: Yes, although Donut/Nougat also have some quite specific postprocessing: https://github.com/huggingface/transformers/blob/main/docs/source/en/model_doc/donut.md#inference-examples

**Comment by NielsRogge** [comment]
@merveenoyan yes the outputs are as expected, they output what you would also get by running processor.batch_decode on the generated token IDs of these models.

The reason KOSMOS-2 outputs a set of random tokens first is because it also outputs tokens for the image patches, which can be discarded.

Similarly, Fuyu-8B first outputs a large set of |SPEAKER| tokens before the actual answer.

The reason you're getting an error for Llava-1.5B is because you're not including an <image> token in the prompt, it's recommended to leverage this prompt template: `"USER: <image>\nWhat are these?\nASSISTANT:"`.

**Comment by NielsRogge** [comment]
@amyeroberts I'd like to assign you for review, but the CI keeps failing with unrelated changes (cc @ydshieh):
```python
TypeError: snapshot_download() got an unexpected keyword argument '_from_pipeline'
```
I've rebased with main already several times. Other than that, the PR is ready for review.

**Comment by amyeroberts** [comment]
Thanks for working on this! 

Just skimming the PR, there's code which shouldn't be be pushed - please make sure to look at the diff in the PR to catch this before asking for review. 

Once it's ready for review, I'll give a more in-depth look. At a high-level, I'm a bit concerned about the addition of the auto model, when from the preprocessing and processing steps there isn't complete unity in the inputs and outputs e.g. adding `decoder_input_ids` for donut.

As discussed a week or two ago, it would be good to have a method for all the model's processors which handle the postprocessing of generation outputs for these models. cc @molbap.

**Comment by NielsRogge** [comment]
Removed the script, PR should be ready for review now

**Review by amyeroberts — COMMENTED** [review]
Thanks for working on this! 

My main comment from before still applies here: I'm concenred that bundling these models together under this auto class doesn't make sense, in particular including the `vision-encoder-deocder` model type as it's clear the inputs and their preparation aren't consistent, resulting in a lot of additional code needed within the pipeline itself. The pipeline shouldn't need to know about the model it's loading

File: src/transformers/pipelines/image_text_to_text.py [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+
+        model_type = self.model.config.model_type
+
+        kwargs = {}
+
+        if model_type == "pix2struct":
+            kwargs = {"add_special_tokens": False}
+
+        if model_type == "idefics":
+            model_inputs = self.processor(text, return_tensors=self.framework, **kwargs)
```
[the remark this answers is not in the document]
Comment (merveenoyan) [reply src/transformers/pipelines/image_text_to_text.py:137]: It takes a sequence (in a list) of processed image and text [apparently](https://huggingface.co/docs/transformers/en/model_doc/idefics#transformers.IdeficsProcessor.__call__.example)

File: tests/models/git/test_modeling_git.py [hunk 4]
```diff
@@ -398,9 +398,7 @@ class GitModelTest(ModelTesterMixin, GenerationTesterMixin, PipelineTesterMixin,
     all_model_classes = (GitModel, GitForCausalLM) if is_torch_available() else ()
     all_generative_model_classes = (GitForCausalLM,) if is_torch_available() else ()
     pipeline_model_mapping = (
-        {"feature-extraction": GitModel, "image-to-text": GitForCausalLM, "text-generation": GitForCausalLM}
-        if is_torch_available()
-        else {}
+        {"feature-extraction": GitModel, "image-to-text": GitForCausalLM} if is_torch_available() else {}
```
[the remark this answers is not in the document]
Comment (NielsRogge) [reply tests/models/git/test_modeling_git.py:401]: Cause Git is inherently a multimodal model, nobody would use it for text generation tasks (although it is compatible with it).

**Comment by MoritzLaurer** [comment]
Is this PR still planned? It would be useful, see e.g. these internal threads ([thread1](https://huggingface.slack.com/archives/C01EDEAV0NT/p1720004277653459?thread_ts=1711469424.532149&cid=C01EDEAV0NT), [thread2](https://huggingface.slack.com/archives/C01N44FJDHT/p1719998786039139?thread_ts=1719997770.767279&cid=C01N44FJDHT)) @NielsRogge

**Comment by NielsRogge** [comment]
Yes it's still planned, I just don't have the bandwidth to work on it now. I hope @molbap @zucchini-nlp have.

Although it would require to have the refactor of all multimodal processors first before this can land.

**Pull request closed without merging**